### PR TITLE
CMake: Add compiler warning about implicit fallthough

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

**Comment by chriseth** [comment]
The specific cases are being fixed in https://github.com/ethereum/solidity/pull/2934

**Comment by chriseth** [comment]
But I guess it is good to have the general warning.

**Comment by axic** [comment]
I'd prefer fixing the cases an not using this annotation, haven't looked at the rest of the PR.

**Comment by chfast** [comment]
The other part enforces the warning (it was not enabled in clang and some later version of GCC 7).

Do you want to keep the portable macro `FALLTHOUGH`?

**Comment by chfast** [comment]
Just a notice, the `[[fallthrough]]` the standard annotation is C++17.

**Comment by axic** [comment]
I mean our goal should be not having fallthroughs at all and by having the macro means we encourage it, right?

Lets wait a few days to see if we can finish and merge #2934.

**Comment by chfast** [comment]
I don't care. I'm happy to remove everything except CMake part from this PR.

BTW
```
case A:
    ...
    goto case B;
case B:
```

**Comment by axic** [comment]
Is there a compiler switch to treat goto as error? :)

I like goto in specific instances, but likely we do not have any need for that in Solidity.

**Comment by chfast** [comment]
Probably not, but there might be a rule in clang-tidy.

File: cmake/EthCompilerSettings.cmake [hunk 1]
```diff
@@ -14,18 +14,15 @@
...
-if (have_stack_protector_strong)
-	add_compile_options(-fstack-protector-strong)
-else()
-	check_cxx_compiler_flag(-fstack-protector have_stack_protector)
-	if(have_stack_protector)
-		add_compile_options(-fstack-protector)
-	endif()
+eth_add_cxx_compiler_flag_if_supported(-fstack-protector-strong have_flag)
+if(NOT have_flag)
+	eth_add_cxx_compiler_flag_if_supported(-fstack-protector)
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake]: Do you want to create a PR for the cmake and the stack protector change (and the URL stuff above)? That is something we should merge now.

File: cmake/EthCheckCXXCompilerFlag.cmake [hunk 2]
```diff
@@ -0,0 +1,16 @@
+include(CheckCXXCompilerFlag)
```
Comment (axic) [inline cmake/EthCheckCXXCompilerFlag.cmake:1]: Can you add a comment to explain how / why is this better than the other `CheckCXXCompilerFlag`?

File: cmake/EthCompilerSettings.cmake (hunk 1 above)
Comment (chfast) [reply cmake/EthCompilerSettings.cmake]: "stack protector" is only a refactor (I created a helper function to handle flags). But you cannot merge "fallthough warning" as solidity will not compile.

Comment (axic) [reply cmake/EthCompilerSettings.cmake]: I only meant the stack protector part and the helper to be merged.

**Comment by chriseth** [comment]
Merged the other pull request, please rebase.

**Comment by chfast** [comment]
I only left the CMake changes.

File: cmake/EthCheckCXXCompilerFlag.cmake (hunk 2 above)
Comment (axic) [reply cmake/EthCheckCXXCompilerFlag.cmake:1]: @chfast can you add a comment here?

Comment (axic) [reply cmake/EthCheckCXXCompilerFlag.cmake:1]: Ah its a combination of `check_cxx_compiler_flag` and `add_compile_options`

**Review by axic — APPROVED** [review]

File: cmake/EthCompilerSettings.cmake [hunk 3]
```diff
@@ -14,18 +14,15 @@
...
-check_cxx_compiler_flag(-fstack-protector-strong have_stack_protector_strong)
-if (have_stack_protector_strong)
-	add_compile_options(-fstack-protector-strong)
-else()
-	check_cxx_compiler_flag(-fstack-protector have_stack_protector)
-	if(have_stack_protector)
-		add_compile_options(-fstack-protector)
-	endif()
+eth_add_cxx_compiler_flag_if_supported(-fstack-protector-strong have_flag)
+if(NOT have_flag)
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake]: I think `have_flag` might be too generic here

**Comment by chfast** [comment]
Addressed your comments, @axic. I will merge when CI is green.

**Pull request merged**